_specifications/lsp/3.18/language/inlineCompletion.md

```diff
@@ -2,7 +2,25 @@
 
 > *Since version 3.18.0*
 
-The inline completion request is sent from the client to the server to compute inline completions for a given text document either explicitly by a user gesture or implicitly when typing.
+The inline completion request is sent from the client to the server to compute
+inline completions for a given text document either explicitly by a user gesture
+or implicitly when typing. Inline completion items are presented directly in the
+editor with the user's text. The portion of the inline completion item's text
+that isn't in the document receives a slightly different UI treatment e.g. lower
+contrast. With only one item being displayed at a time, the server may use this
+feature for visible edits where it is relatively confident the user would like
+this edit in particular. The context property also provides allowance for this
+feature to work *with* completions presented in the IntelliSense interface.
+
+When multiple inline completion items are returned, the client may decide
+whether the user can cycle through them or if they, along with their
+`filterText`, are merely for filtering if the user continues to type without yet
+accepting the inline completion item.
+
+Clients may choose to send information about the user's current completion
+selection via context if completions are visible at the same time. In this case,
+returned inline completions should extend the text of the provided completion.
```

Comment:
In what way is this different from a client request `textDocument/completion` and rendering the results as "inline" anyway?

What does it matter to a server whether the completions are rendered inline or with a traditional widget box?

Reply:
I rephrased this since @mfussenegger is correct that we shouldn't mention any required UI implementations in the specification.

```
The inline completion request is sent from the client to the server to compute inline completions for a given text document either explicitly by a user gesture or implicitly when typing. 

Inline completion items usually complete bigger portions of text (e.g. whole methods, ...) and in contrast to completions items can complete code that might be syntactically or semantically incorrect. 

Due to this inline completion items are usually not suited to be presented in normal code completion widgets like a list of items. One possible approach can be to present the information inline in the editor with lower contrast.

When multiple inline completion items are returned, the client may decide whether the user can cycle through them or if they, along with their `filterText`, are merely for filtering if the user continues to type without yet accepting the inline completion item.

Clients may choose to send information about the user's current completion selection via context if completions are visible at the same time. In this case, returned inline completions should extend the text of the provided completion.
```

Reply:
Phrasing looks good to me.